qualang_tools/control_panel/video_mode/data_acquirers.py

```diff
@@ -0,0 +1,360 @@
+        )
+        for axis in [self.x_axis, self.y_axis]:
+            label = axis.label or axis.name
+            self.data_array.coords[axis.name].attrs.update({"units": axis.units, "long_name": label})
+        logging.debug("DataGenerator initialized with initial data")
+
+    def update_attrs(self, attrs: List[Dict[str, Any]]):
+        """Update the attributes of the data array.
+
+        This method updates the attributes of the data array with the provided attributes.
+        This method is invoked from the Dash app when the user updates the parameters in the control panel.
+
+        Args:
+            attrs: The attributes to update.
+        """
+        for attr in attrs:
+            setattr(attr["obj"], attr["key"], attr["new"])
+        pass
+
+    @abstractmethod
+    def acquire_data(self) -> np.ndarray:
+        """Acquire data from the device.
+
+        This method must be implemented by subclasses to provide the actual data acquisition logic.
+        """
+        pass
+
+    def update_data(self):
+        """Update the data array with the new data.
+
+        This method acquires new data from the device and updates the data array.
+        It also performs a rolling average of the data to reduce noise.
+        """
+        new_data = self.acquire_data()
+        self.num_acquisitions += 1
+
+        if new_data.shape != self.data_array.values.shape:
+            self.data_history.clear()
+
+        self.data_history.append(new_data)
+
+        if len(self.data_history) > self.num_averages:
+            self.data_history.pop(0)
+
+        averaged_data = np.mean(self.data_history, axis=0)
+
+        self.data_array = xr.DataArray(
+            averaged_data,
+            coords=[
+                (self.x_axis.name, self.x_axis.sweep_values_with_offset),
+                (self.y_axis.name, self.y_axis.sweep_values_with_offset),
+            ],
+            attrs=self.data_array.attrs,  # Preserve original attributes like units
+        )
+        for axis in [self.x_axis, self.y_axis]:
+            label = axis.label or axis.name
+            self.data_array.coords[axis.name].attrs.update({"units": axis.units, "long_name": label})
+        mean_abs_data = np.mean(np.abs(averaged_data))
+        logging.debug(f"Data acquired with shape: {self.data_array.shape}, mean(abs(data)) = {mean_abs_data}")
+        return self.data_array
+
+
+class RandomDataAcquirer(BaseDataAcquirer):
+    """Data acquirer that acquires random data."""
+
+    def acquire_data(self):
+        """Acquire random data.
+
+        This method acquires random data from the simulated device.
+        """
+        sleep(1)
+        results = np.random.rand(self.x_axis.points, self.y_axis.points)
+        return results
+
+
+class OPXDataAcquirer(BaseDataAcquirer):
+    """Data acquirer for OPX devices.
+
+    This class is responsible for acquiring data from OPX devices.
+
+    Args:
+        qm: The QuantumMachine instance.
```

Comment:
The docstring must be with qmm and qua_config

Reply:
Fixed